### PR TITLE
[Concat] Refactor kernel functions

### LINKED ISSUE #439: Kernel refactor for `Concat`

### PR DESCRIPTION
**Pull request by virena**
**Description of changes:**

This PR removes kernel functions from `concat.h` and `concat.cc` into separate `concat_kernels.h`, `concat_kernels.cu`, and `concat_kernels.cpp`.

**Related Issues:**

Linked Issues:
- Issue #303

Issues closed by this PR:
- Closes #439 

**Before merging:**

- [x] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules? **n/a**

### CODE REVIEW

File: src/ops/kernels/concat_kernels.cpp [hunk 1]
```diff
@@ -25,6 +25,74 @@ using Legion::coord_t;
...
 using Legion::Rect;
 
+namespace Kernels {
+namespace Concat {
+
+void init_meta(ConcatMeta *m, int legion_axis) {
+  m->legion_axis = legion_axis;
+}
+
+/*static*/
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cpp:35]:
```suggestion
```

File: src/ops/kernels/concat_kernels.cpp [hunk 2]
```diff
@@ -25,6 +25,74 @@ using Legion::coord_t;
...
+    float elapsed = 0;
+    checkCUDA(hipEventElapsedTime(&elapsed, t_start, t_end));
+    printf("[%s] forward time = %.4f ms\n", m->op_name, elapsed);
+    hipEventDestroy(t_start);
+    hipEventDestroy(t_end);
+  }
+}
+
+/*static*/
+void Concat::backward_kernel_wrapper(ConcatMeta const *m,
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cpp:68]:
```suggestion
void backward_kernel_wrapper(ConcatMeta const *m,
```

File: src/ops/kernels/concat_kernels.cu [hunk 3]
```diff
@@ -13,7 +13,7 @@
  * limitations under the License.
  */
 
-#include "flexflow/ops/concat.h"
+#include "flexflow/ops/kernels/concat_kernels.h"
 #include "flexflow/utils/cuda_helper.h"
 #include "flexflow/utils/hash_utils.h"
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cu:18]: Why necessary?

File: src/ops/kernels/concat_kernels.cu [hunk 4]
```diff
@@ -24,6 +24,74 @@ using Legion::coord_t;
 using Legion::Domain;
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cu:24]: Why necessary?

File: src/ops/kernels/concat_kernels.cpp [hunk 5]
```diff
@@ -13,7 +13,7 @@
  * limitations under the License.
  */
 
-#include "flexflow/ops/concat.h"
+#include "flexflow/ops/kernels/concat_kernels.h"
 #include "flexflow/utils/hash_utils.h"
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cpp:17]: Why necessary?

File: src/ops/kernels/concat_kernels.cpp [hunk 6]
```diff
@@ -25,6 +25,74 @@ using Legion::coord_t;
 using Legion::Domain;
```
Comment (lockshaw) [inline src/ops/kernels/concat_kernels.cpp:25]: Why necessary?

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: src/ops/kernels/concat_kernels.cu (hunk 3 above)
Comment (virena) [reply src/ops/kernels/concat_kernels.cu:18]: This is for the kernel function implementations, so we should include that header file. However, we don't need any of the methods in `concat.h`.

File: src/ops/kernels/concat_kernels.cu (hunk 4 above)
Comment (virena) [reply src/ops/kernels/concat_kernels.cu:24]: Removed.

Comment (virena) [reply src/ops/kernels/concat_kernels.cu:24]: Removed.

File: src/ops/kernels/concat_kernels.cpp (hunk 5 above)
Comment (virena) [reply src/ops/kernels/concat_kernels.cpp:17]: Same as above

File: src/ops/kernels/concat_kernels.cu (hunk 3 above)
Comment (lockshaw) [reply src/ops/kernels/concat_kernels.cu:18]: I think I meant the `hash_utils` here :slightly_smiling_face:

Comment (virena) [reply src/ops/kernels/concat_kernels.cu]: Removed.

**Review by lockshaw — APPROVED** [review]

**Pull request merged**